### PR TITLE
Issue4 #4 

### LINKED ISSUE #4: test x arrays container
test x arrays container for weatherdata to store meta-datainformation, coordinate, units, variable labels.

### PR DESCRIPTION
**Pull request by mlabadie**
# Test weatherdata with xarray.data instead of a panda dataframe 

Now request of ipm weatherdata througth weatherdata module  return a pandadataframe. 

**Objective of this PR is to return a dataset with metainformation**
For this we decided used xarray module to allows us store meta information in attributs

Currently we have managed to return to xarray dataset from the query with all the metainformation 
- for the dataset the name of the weatheresource, the station and its longitude and lattitude, the start and end dates, the length of the dataset and the quality control information for each parameter (weather variable)
- for each variable we also return in attribute the id code, the full name of the variable, the description and the unit
- for the coordinates a description of each coordinate

Now the aim are to retour a dataset with information of all station in a single dataset each weatherressource and forecast

### CODE REVIEW

**Comment by mlabadie** [comment]
Bug in time coordinate is not considering like a datetime type but a type object

**Comment by mlabadie** [comment]
# small point on the package
Now, we can query on a list of station_id for weatheradapter and a list of latitude, longitude and altitude for forcast in data function

## Todo:
- tests for all the functions in the package
- documentations

**Comment by mlabadie** [comment]
I I did some tests for weatherdata ipm. All the tests passed except for 4, which is normal. The tests test the different resources and the production of an xarray dataset or a json. 
I have also created a notes file in which I have listed the different problems in the execution of the different resources. 
In ipm I created a cache system (not definitive but it allows me to avoid restarting all the requests each time and a pseudo log where I print the different requests for the different stations.

File: README.md [hunk 1]
```diff
@@ -15,20 +15,26 @@ An open source library for data structure for handling weather data
...
 git clone https://github.com/H2020-IPM-openalea/weatherdata.git
+cd weatherdata
+python setup.py install
+
 
-python setup.py develop
 ```
 
 ## Requierements
     * python2.7
```
Comment (pradal) [inline README.md:33]: Python or python 2.7

File: src/weatherdata/ipm.py [hunk 2]
```diff
@@ -10,9 +10,16 @@
...
 import datetime
+import xarray as xr
+import numpy as np
+import os
+import json
 
 from agroservices import IPM
 
+usecache=True
+pathcache=r"C:\Users\mlabadie\Documents\GitHub\ipm\weatherdata\cache"
```
Comment (pradal) [inline src/weatherdata/ipm.py]: WARNING: No explicit path...

Comment (pradal) [reply src/weatherdata/ipm.py]: Manage the cache automatically

File: src/weatherdata/ipm.py [hunk 3]
```diff
@@ -54,18 +61,23 @@ def station_ids(self):
         coord = dict()
 
         for names in values:
-            if 'features' in values[names]:
+            if 'features' in values[names] and values[names]['features']!=[]:
```
Comment (pradal) [inline src/weatherdata/ipm.py]: Difficult to read:
Use `feature = values[names]['features']` instead

File: src/weatherdata/ipm.py [hunk 4]
```diff
@@ -54,18 +61,23 @@ def station_ids(self):
...
                 station_ids[names]=[values[names]['features'][item]['properties'] for item in range(len(values[names]['features']))]
                 coord[names]=[values[names]['features'][item]['geometry'] for item in range(len(values[names]['features']))]
             else:
                 station_ids[names]= 'no stations for this ressources'
                 coord[names]= 'no stations for this ressources'
+        
+        if station_ids[self.name]=='no stations for this ressources':
```
Comment (pradal) [inline src/weatherdata/ipm.py]: Warning: do not put the message in the code.
Perhaps, one solution would be to manage a code associated with the message in a separate module?

File: src/weatherdata/ipm.py [hunk 5]
```diff
@@ -131,108 +143,175 @@ def check_forecast_endpoint(self):
...
-            latitude: (double) WGS84 Decimal degrees
-            longitude: (double) WGS84 Decimal degrees
+            altitude: (list) only for Met Norway Locationforecast WGS84 Decimal degrees
+            latitude: (list) WGS84 Decimal degrees
+            longitude: (list) WGS84 Decimal degrees
         
         Returns:
         --------
-            return a dataframe (ViewDataFrame=True) or json format (ViewDataFrame=False) 
+            return a dataset (format='ds') or json format (format='json')
```
Comment (pradal) [inline src/weatherdata/ipm.py:172]: dataset -> xarray dataset

File: src/weatherdata/ipm.py [hunk 6]
```diff
@@ -131,108 +143,175 @@ def check_forecast_endpoint(self):
...
-                data = {str(var): vals for var, vals in zip(response['weatherParameters'], zip(*response['locationWeatherData'][0]['data']))}
-                df = pandas.DataFrame(data)
-                df.index = daterange
-                # TODO : get all what is needed for intantiating a WeatherData object (meta, units, ...) and retrun it
-                return df
+            interval = pandas.Timedelta(times.freq).seconds
+            
+            responses = []
+            for station in stationId:
+                print('start connecting to station', station)
```
Comment (pradal) [inline src/weatherdata/ipm.py]: use a logger instead

File: src/weatherdata/ipm.py [hunk 7]
```diff
@@ -131,108 +143,175 @@ def check_forecast_endpoint(self):
...
+            times = pandas.date_range(
+                start=responses[0]['timeStart'], 
+                end=responses[0]['timeEnd'], 
+                freq="H",
+                name="time")
+
+
+        if format == 'ds':
+            #data conversion in numpy array
+            data= [np.array(responses[el]['locationWeatherData'][0]['data']) for el in range(len(responses))]
```
Comment (pradal) [inline src/weatherdata/ipm.py]: This code needs to be simplified:

```
data= [np.array(responses[el]['locationWeatherData'][0]['data']) for el in range(len(responses))]
```
to 
```
data= [np.array(wstation['locationWeatherData'][0]['data']) for wstation in responses]
```

File: src/weatherdata/ipm.py [hunk 8]
```diff
@@ -131,108 +143,175 @@ def check_forecast_endpoint(self):
...
+            ds=xr.combine_by_coords(list_ds)
+            
+            #coordinates attributes
+            if stationId is not None:
+                ds.coords['location'].attrs['name']= 'WeatherStationId'
             else:
-                return response
-
-    
+                ds.coords['location'].attrs['name']='[latitude,longitude]'
```
Comment (pradal) [inline src/weatherdata/ipm.py:279]: This line is strange.
Document it?

File: src/weatherdata/xarray_test.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
```
Comment (pradal) [inline src/weatherdata/xarray_test.py:1]: You have to add a test!
Name it test_xarray.

**Pull request merged**